### DISCUSSION TITLE
Return Trace ID in HTTP headers

### DISCUSSION THREAD

**Discussion by amexboy** [open-telemetry/opentelemetry-java-instrumentation#8175] [Q&A]
I couldn't find any information if this is by design or if I miss-configured something. 

I just implemented OpenTelemetry on a new Spring Boot application. I can see tracing works as I can see the requests in Jaeger and also in the console from the logging exporter. However, I don't get the trace-id back in the `curl` request. 

I would like to have this returned for the original request. This not only facilitate testing, it will also make it easy for clients to report problems. This doesn't seem to be the default behavior. Looking through the code for the code for the instrumenter (`SpringWebMvcTelemetry` and related classes), it only extracts information from the headers and doesn't seem to add anything once the request is done processing.  

Is this already supported? Is there any reason this is not implemented or should not be implemented?

P.S:  I can easily implement a Spring Filter to get the current context and return the relevant headers. Or perhaps use create `OperationListener` to copy the context to the HTTP response headers.

**Comment by jkwatson** [comment]
moving to the instrumentation repository

**Comment by ddark008** [comment]
You can modify response by [HttpServerResponseCustomizer](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/5afe4d203524cea9907e4e5393feb74f52844949/javaagent-extension-api/src/main/java/io/opentelemetry/javaagent/bootstrap/http/HttpServerResponseCustomizer.java)
Something like this:
```java
@AutoService(HttpServerResponseCustomizer.class)
public class TraceIdHttpServerResponseCustomizer implements HttpServerResponseCustomizer {
    @Override
    public <T> void customize(Context serverContext, T response, HttpServerResponseMutator<T> responseMutator) {
        SpanContext spanContext = Span.fromContext(serverContext).getSpanContext();
        if (!spanContext.isValid()) {
            return;
        }
        responseMutator.appendHeader(response, "X-Trace-Id", spanContext.getTraceId());
    }
}
```

**Comment by dalbani** [comment]
Do I misunderstand or this is actually what the [(draft) standard `traceresponse` header](https://github.com/w3c/trace-context/blob/main/spec/21-http_response_header_format.md) is about?

**Reply by trask** [reply]
I believe so, there was some initial work in getting this into OpenTelemetry specification, but it needs to be picked back up still: https://github.com/open-telemetry/opentelemetry-specification/pull/3825

**Comment by dalbani** [comment]
Note that Spring Boot 3.5.0-M2 is now capable of including an `X-Trace-Id` header in the response:
- https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-3.5.0-M2-Release-Notes#write-trace-ids-in-http-response-headers
- https://github.com/spring-projects/spring-boot/issues/40857.
